tools/ci_connector_ops/ci_connector_ops/pipelines/actions/environments.py

```diff
@@ -101,45 +98,74 @@ async def with_python_package(
     container = python_environment.with_mounted_directory("/" + package_source_code_path, package_source_code_directory).with_workdir(
         "/" + package_source_code_path
     )
+    return container
+
+
+async def with_installed_python_package(
+    context: ConnectorTestContext,
+    python_environment: Container,
+    package_source_code_path: str,
+    additional_dependency_groups: Optional[List] = None,
+    exclude: Optional[List] = None,
+) -> Container:
+    """Installs a python package in a python environment container.
+
+    Args:
+        context (ConnectorTestContext): The current test context, providing the repository directory from which the python sources will be pulled.
+        python_environment (Container): An existing python environment in which the package will be installed.
+        package_source_code_path (str): The local path to the package source code.
+        additional_dependency_groups (Optional[List]): extra_requires dependency of setup.py to install. Defaults to None.
+        exclude (Optional[List]): A list of file or directory to exclude from the python package source code.
+
+    Returns:
+        Container: A python environment container with the python package installed.
+    """
 
-    if install:
-        if requirements_txt := await get_file_contents(container, "requirements.txt"):
-            for line in requirements_txt.split("\n"):
-                if line.startswith("-e ."):
-                    local_dependency_path = package_source_code_path + "/" + line[3:]
-                    container = container.with_mounted_directory(
-                        "/" + local_dependency_path, context.get_repo_dir(local_dependency_path, exclude=DEFAULT_PYTHON_EXCLUDE)
-                    )
-            container = container.with_exec(INSTALL_LOCAL_REQUIREMENTS_CMD)
+    container = with_python_package(context, python_environment, package_source_code_path, exclude=exclude)
+    if requirements_txt := await get_file_contents(container, "requirements.txt"):
```

Comment:
I totally get it, but the function parameters need to be hashable to be cached right? The `Container` object is not hashable.

Reply:
> The Container object is not hashable.

If it's not, it can be made to be. I'll make an issue for it. 

A `@chachedproperty` like thing is simpler. If in a class method (e.g., `Connector` instance):

```python
@cachedproperty
async def with_installed_python_package(self):
    # same as what you have already
    return container
```

Which is basically doing this:
```python
async def with_installed_python_package(self):
    if not self._installed_package:
        # do the thing
        self._installed_package = container
    return self._installed_package
```

Since this should be the same per connector, a single object instance that has methods for connector wide dependencies like this can make things simpler.